### PR TITLE
Add code to parse the policy from a global array

### PR DESCRIPTION
**Pull request by jaskarankhurana**
The array represents the files where each entry is a rule.
Add parsing, printing and cleanup for the policy.

Signed-off-by: Jaskaran Khurana <jaskarankhurana@linux.microsoft.com>

### CODE REVIEW

File: security/ipe/ipe-core.c [hunk 1]
```diff
@@ -11,6 +11,15 @@
 #include "ipe-policy.h"
 #include "ipe-audit.h"
 
+static struct ipe_property properties[] = {
```
Comment (bowerscd) [inline security/ipe/ipe-core.c]: why was this moved to core?

File: security/ipe/ipe-parse.c [hunk 2]
```diff
@@ -0,0 +1,233 @@
+// SPDX-License-Identifier: GPL-2.0
+#include <linux/types.h>
+#include <linux/binfmts.h>
+#include <linux/parser.h>
+#include <linux/slab.h>
+#include "ipe.h"
+#include "ipe-property.h"
+#include "ipe-parse.h"
+
+static char *rule_table[] = {
```
Comment (bowerscd) [inline security/ipe/ipe-parse.c]: I know this won't be the way it is for an external thing, but `const char*`

File: security/ipe/ipe-parse.c [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+static int ipe_get_operation_table_mapping(char *operation_name,
+					   enum ipe_hook *op)
+{
+	size_t i;
+
+	if (operation_name == NULL || op == NULL)
+		return -EINVAL;
+
+	for (i = 0; i < sizeof(ops_table); i++) {
+		if (strcmp(ops_table[i].operation_name, operation_name) == 0) {
```
Comment (bowerscd) [inline security/ipe/ipe-parse.c]: this data won't necessarily be trusted. strncmp instead? as-is it's trusted, but for future plans it will not be

File: security/ipe/ipe-parse.c [hunk 4]
```diff
@@ -0,0 +1,233 @@
...
+			pr_info("IPE rule %lu {", j);
+			pr_info("is_allow_rule=%s,",
+				lsm_rules[i].ipe_rules[j].is_allow_rule ?
+				"true" : "false");
+			print_sub_rule(&lsm_rules[i].ipe_rules[j]);
+			pr_info ("}");
+		}
+		pr_info("}");
+	}
+}
```
Comment (bowerscd) [inline security/ipe/ipe-parse.c]: move to utils

File: security/ipe/ipe-policy.c [hunk 5]
```diff
@@ -1,65 +1,11 @@
...
-		IPE_INIT_DENY_RULE(
-			IPE_INIT_SUBRULE(IPE_PROPERTY_FILEPATH, "/usr/bin/*"))),
-	IPE_INIT_RULE_TABLE(
-		IPE_INIT_ALLOW_RULE(IPE_INIT_SUBRULE(IPE_PROPERTY_DMVERITY,
-						     IPE_PROPERTY_VALUE_TRUE)),
-		IPE_INIT_ALLOW_RULE(
-			IPE_INIT_SUBRULE(IPE_PROPERTY_FILEPATH, "/root/*")),
-		IPE_INIT_ALLOW_RULE(IPE_INIT_SUBRULE(IPE_PROPERTY_FILEPATH,
-						     "/lib/*.so*"))));
+struct ipe_rule_table lsm_rules[IPE_HOOK_MAX];
```
Comment (bowerscd) [inline security/ipe/ipe-policy.c]: do we think this file is necessary anymore?

File: security/ipe/ipe-parse.h [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+#include "ipe.h"
+#include "ipe-hooks.h"
+#include "ipe-property.h"
+
+#define MAX_RULES 256
+
+#define ALLOW "allow"
+#define DENY "deny"
+
+enum {
```
Comment (bowerscd) [inline security/ipe/ipe-parse.h]: no type name for enum?

prefix enum with ipe_* to avoid possible-namespace collisions?

File: security/ipe/ipe-parse.c (hunk 4 above)
Comment (bowerscd) [reply security/ipe/ipe-parse.c]: (print functions)

File: security/ipe/ipe-core.c (hunk 1 above)
Comment (jaskarankhurana) [reply security/ipe/ipe-core.c]: static structures should not be in .h so will move to ipe-core.c

File: security/ipe/ipe-parse.c (hunk 2 above)
Comment (jaskarankhurana) [reply security/ipe/ipe-parse.c]: Yes.

File: security/ipe/ipe-parse.c (hunk 3 above)
Comment (jaskarankhurana) [reply security/ipe/ipe-parse.c]: Yes.

Comment (jaskarankhurana) [reply security/ipe/ipe-parse.c]: yep also strtok will make sure its null terminated so we should be fine here.

File: security/ipe/ipe-policy.c (hunk 5 above)
Comment (bowerscd) [reply security/ipe/ipe-policy.c]: this still needs to be removed; otherwise I approve.
